### ISSUE TITLE
Unlimited Voting Through API

### ISSUE THREAD

**Issue by BrokenEagle** [r888888888/danbooru#2719]
Whenever attempting to vote on a post through the API, it always succeeds  `{"success": true}`, regardless the number of times a vote has been added from an account.

Check http://testbooru.donmai.us/posts/1 for an example of this.  All 10 of those votes came from one account.

Whenever attempting to unvote on a post through the API it always fails `{"success":false,"reason":"access denied"}`.

Whenever attempting to vote on a comment through the API, it always fails `{"success":false,"reason":"access denied"}`.

Whenever attempting to unvote on a comment through the API, it always generates a "Missing Template Exemption", just like issue #2700.
`{"success": false, "message": "Missing template comments/unvote, application/unvote with {:locale=\u003e[:en], :formats=\u003e[:json], :variants=\u003e[], :handlers=\u003e[:erb, :builder, :raw, :ruby, :coffee]}. Searched in:\n  * \"/var/www/danbooru2/releases/20161006212647/app/views\"\n"}`

I've validated that all of the above also holds true on Danbooru, so there are also a few superfluous votes on http://danbooru.donmai.us/posts/1.

**Comment by evazion** [comment]
>Whenever attempting to vote on a post through the API, it always succeeds {"success": true}, regardless the number of times a vote has been added from an account.

How are you sending the request? I can't reproduce this. Tried it like this:

```
USER=evazion
KEY=GnPkLxcASJhWemtvlpGjhwe1sk8lcttX91Z7Q96vqOg
SITE="http://testbooru.donmai.us"

$ curl -u "$USER:KEY" -X POST "$SITE/posts/1/votes.json" -d 'score=up'
  {"success": true}
$ curl -u "$USER:KEY" -X POST "$SITE/posts/1/votes.json" -d 'score=up'
  {"success": false, "reason": "You have already voted for this post"}
$ curl -u "$USER:KEY" -X PUT "$SITE/posts/1/unvote.json"
{"success": false, "message": "Missing template posts/unvote, application/unvote with {:locale=\u003e[:en], :formats=\u003e[:json], :variants=\u003e[], :handlers=\u003e[:erb, :builder, :raw, :ruby, :coffee]}. Searched in:\n  * \"/var/www/danbooru2/releases/20161006212647/app/views\"\n"}
```

I was able to vote (which should not be possible for member-level users, so something weird definitely is going on) but not able to vote twice. 

> Whenever attempting to unvote on a post through the API it always fails {"success":false,"reason":"access denied"}.

This is strange. Unvoting via `PUT /posts/1/unvote.json` succeeds for me despite returning a missing template error, but unvoting via `DELETE /posts/1/votes` fails with a 404, even though that's supposed to unvote as well.

**Comment by BrokenEagle** [comment]
Hell if I know what's going on...? :( 

 I have been using HTTPRequester which is an addon for Firefox for most of my testing, but I switched back to Python to test alternatives.

`resp = requests.post('http://testbooru.donmai.us/posts/1/votes.json?username=BrokenEagle98&apikey=W29N5XQa9tLkjM6m_2y5z5XZ-WSgAPJJBNX0MhqMzt8&score=up`

Status Code & Reason: 200 OK
Content: '  {"success": true}\n'

I also switched to using your credentials only to have the same results.  Finally, I moved the "score=up" parameter in the URL to the POST data instead, to still have the same results, so... ¯\_(ツ)_/¯ 

As you can see on http://testbooru.donmai.us/posts/1 , the score is up to 18... they were all a success...

**Comment by evazion** [comment]
Alright, I think I've figured this out. There was a regression in 6d6d00b190dd76aabcfaa54cb10c8252d4e844a1; [before_filter :voter_only](https://github.com/r888888888/danbooru/commit/6d6d00b190dd76aabcfaa54cb10c8252d4e844a1#diff-49b87e0bf84b9c7f23be828712aad007R2) is a no-op there because returning false does not halt the request. The fix is to arrange for a `voter_only` method to be defined [here](https://github.com/r888888888/danbooru/blob/master/app/controllers/application_controller.rb#L136-L145) that properly redirects to the access denied page.

Unlimited voting came from a combination of factors:

1. This bug makes it possible for anyone, including anonymous (not logged in) users, to vote.
2. When anonymous users vote apparently no post vote is recorded, so they can vote repeatedly.
3. You weren't logged in properly - the proper params are `login` and `api_key` - so you were voting as anonymous. When logged in properly you are blocked from voting twice.

**Comment by BrokenEagle** [comment]
>Evazion said:
 You weren't logged in properly - the proper params are login and api_key - so you were voting as anonymous. When logged in properly you are blocked from voting twice.

Ah, that's true... and a bit embarrassing... >_< Still, it was a bit fortuitous that I did mess up I guess, as I was able to find that bug with it...

Although I'm still having the same issues that you noted for the unvote function like in issue #2700.
